### PR TITLE
Add retrievers using the parser-only approach

### PR DESCRIPTION
**Pull request by jdconrad**
This enhancement adds a new abstraction to the `_search` API called "retriever." A retriever is something that returns top hits.

Three retrievers are going to be added initially:
* `standard` - This retriever replaces standard query functionality. It contains sub-elements for `query`, `search_after`, `terminate_after`, `sort_field`, `min_score`, and `collapse`.
* `knn` - This retriever replaces a knn search. It contains sub-elements for `field`, `k`, `num_candidates`, `query_vector`, `query_vector_builder`, and `similarity`.
* `rrf` - This retriever replaces rank for rrf. It contains sub-elements for `retrievers`, `rank_constant`, and `window_size`.

All retrievers have a `filter` element as well. For `rrf` this is propagated to its child retrievers.

Also included is node feature functionality. Each retriever has its own node feature to ensure compatibility with serverless. New retrievers can be easily added by adding a new node feature without having to worry about incompatibilities.

Retrievers are always "extracted" to a `SearchSourceBuilder`. They are simply an abstraction to better show how elements of the `_search` API interact together in an intuitive way. They are never serialized and are ephemerally used only during parsing.

Some examples:
```
GET index/_search
{
    "retriever": {
        "standard": {
            "query": {
                "match": {
                    "message": {
                        "query": "This is a test."    
                    }
                }
            },
            "sort": [ "popularity" ]
        }
    },
    "size": 5,
    "fields": ["message"]
}

GET index/_search
{
    "retriever": {
        "knn": {
            "field": "vector",
            "k": 3,
            "num_candidates": 10,
            "query_vector": [1, 2, 3]
        }
    },
    "fields": ["message"]
}

GET index/_search
{
    "retriever": {
        "rrf": {
            "window_size": 100,
            "retrievers": [
                {
                    "knn": {
                        "field": "vector",
                        "k": 3,
                        "num_candidates": 10,
                        "query_vector": [1, 2, 3]
                    }
                },
                {
                    "standard": {
                        "query": {
                            "match": {
                                "message": {
                                    "query": "This is a test."    
                                }
                            }
                        }
                    }
                }
            ]
        }
    },
    "size": 5,
    "fields": ["message"]
}
```

### CODE REVIEW

**Comment by jdconrad** [comment]
@javanna @mayya-sharipova A note about this review is that node features required a lot of mechanical changes to anything that uses `SearchSourceBuilder` so a large number of lines are dedicated to plumbing this through as `clusterSupportsFeature`. The main changes are going to be in `RetrieverBuilder`, `StandardRetrieverBuilder`, `KnnRetrieverBuilder`, `RRFRetrieverBuilder`, and `SearchSourceBuilder`.

**Review by mark-vieira — APPROVED** [review]

File: server/src/main/java/org/elasticsearch/search/builder/SearchSourceBuilder.java [hunk 1]
```diff
@@ -81,43 +86,44 @@
...
-    public static final ParseField RESCORE_FIELD = new ParseField("rescore");
-    public static final ParseField STATS_FIELD = new ParseField("stats");
-    public static final ParseField EXT_FIELD = new ParseField("ext");
-    public static final ParseField PROFILE_FIELD = new ParseField("profile");
-    public static final ParseField SEARCH_AFTER = new ParseField("search_after");
-    public static final ParseField COLLAPSE = new ParseField("collapse");
-    public static final ParseField SLICE = new ParseField("slice");
-    public static final ParseField POINT_IN_TIME = new ParseField("pit");
-    public static final ParseField RUNTIME_MAPPINGS_FIELD = new ParseField("runtime_mappings");
+    public static final ParseField FROM_FIELD = new ParseField("from"); // override
```
Comment (mayya-sharipova) [inline server/src/main/java/org/elasticsearch/search/builder/SearchSourceBuilder.java:89]: I think we need more explanations for field comments, what is the difference between `override` VS `retriever` vs `global`

**Comment by mayya-sharipova** [comment]
@jdconrad @javanna I briefly studied this PR, but it seems to me it would be nice to break it up into 2 PRs, as these are 2 types independent types of work :
1. Add `clusterSupportsFeature` to `RestSearchAction`
2. Introduce retrievers.


What do you think of this?

**Comment by javanna** [comment]
agreed @mayya-sharipova

**Comment by jdconrad** [comment]
@mayya-sharipova @javanna The clusterSupportsFeature must be introduced with the retrievers together or it will not work because it would then be possible for a cluster to say it supports a feature it does not yet actually support. As an example if node 1 has retrievers and node 2 does not, if we say both nodes do prior to the retrievers added, then this will break. The original PR for clusterSupportsFeature can be viewed as part of the retrievers branch here (https://github.com/elastic/elasticsearch/pull/105040).

**Comment by jdconrad** [comment]
I'm going to close this PR and re-open with a clean new PR after splitting off the plumbing code.

**Pull request closed without merging**